src/sources/dnstap/parser.rs

```diff
@@ -0,0 +1,1053 @@
+use crate::{
+    event::{LogEvent, PathComponent, Value},
+    Error, Result,
+};
+use bytes::Bytes;
+use prost::Message;
+use snafu::Snafu;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::{convert::TryInto, fmt::Debug};
+use trust_dns_proto::{
+    rr::domain::Name,
+    serialize::binary::{BinDecodable, BinDecoder},
+};
+mod dnstap_proto {
+    include!(concat!(env!("OUT_DIR"), "/dnstap.rs"));
+}
+
+use dnstap_proto::{
+    dnstap::Type as DnstapDataType, message::Type as DnstapMessageType, Dnstap,
+    Message as DnstapMessage, SocketFamily, SocketProtocol,
+};
+
+use super::dns_message::{
+    DnsRecord, EdnsOptionEntry, OptPseudoSection, QueryHeader, QueryQuestion, UpdateHeader,
+    ZoneInfo,
+};
+use super::dns_message_parser::DnsMessageParser;
+use super::schema::DnstapEventSchema;
+
+#[derive(Debug, Snafu)]
+enum DnstapParserError {
+    #[snafu(display("Unsupported DNSTap message type: {}", "dnstap_message_type_id"))]
+    UnsupportedDnstapMessageTypeError { dnstap_message_type_id: i32 },
+}
+
+pub struct DnstapParser<'a> {
+    event_schema: &'a DnstapEventSchema,
+    parent_key_path: Vec<PathComponent>,
+    log_event: &'a mut LogEvent,
+}
+
+impl<'a> DnstapParser<'a> {
+    pub fn new(event_schema: &'a DnstapEventSchema, log_event: &'a mut LogEvent) -> Self {
+        Self {
+            event_schema,
+            parent_key_path: Vec::new(),
+            log_event,
+        }
+    }
+
+    fn insert<V>(&mut self, key: &'static str, value: V) -> Option<Value>
+    where
+        V: Into<Value> + Debug,
+    {
+        let mut node_path = self.parent_key_path.clone();
+        node_path.push(PathComponent::Key(key.to_string()));
+        self.log_event.insert_path(node_path, value)
```

Comment:
Calling `log_event.insert(key, value)` would certainly work (as that's exactly what we did originally), but it'd be much slower than calling `log_event.insert_path(node_path, value)` (yeah, we were surprised too when we found it). Accordingly to our investigation, the performance issue can be rooted to `PathIter`, of which the current implementation is kinda inefficient.

Reply:
Interesting! This is a usefu note.

cc/ @blt you might be interested in this: hat `log_event.insert` is much slower than `log_event.insert_path`. I'm guessing this is due to the former parsing the string as a path. This might be affected by the lookup buf work currently going on though (@StephenWakely may be able to confirm).